package.json

```diff
@@ -25,9 +25,29 @@
       "versions": {
         "0.1.0": "provideSnippets"
       }
+    },
+    "autocomplete.driver": {
+      "versions": {
+        "4.0.0": "provideAutocomplete"
+      }
+    }
+  },
+  "consumedServices": {
+    "snippetsResolver": {
+      "description": "Provide custom functions to resolve snippet variables and transforms.",
```

Comment:
Is the idea here that an arbitrary package could allow the snippets package to understand new variables? I’m curious how that would work in practice and whether two packages that didn’t know about each other could gracefully add non-conflicting extensions to the snippet syntax if they weren’t guaranteed to execute in the same order.

(Also kinda curious about the case where those two packages try to add _conflicting_ resolvers, like if they both want to be in control of what `$FOO` resolves to. But I’m not too worried about that in practice.)

Reply:
> Is the idea here that an arbitrary package could allow the snippets package to understand new variables?

Exactly. Sublime Text [also offers this](http://sublimetext.info/docs/en/extensibility/snippets.html). E.g., I'm imagining a package dedicated to "user details" that provides variables like `AUTHOR` and `EMAIL`. This can get complicated quickly, with wanting different values in different contexts etc., so I feel it's best left as a dedicated provider, instead of direct support here. The feature also means users can tweak all existing behaviour to their liking.

> whether two packages that didn’t know about each other could gracefully add non-conflicting extensions to the snippet syntax if they weren’t guaranteed to execute in the same order

I guess we could add a priority, but I feel it's unnecessary. There's only so many ways to interpret a variable. If two packages try to define the same variable, they are either doing the same thing (so should resolve the same way), or have different ideas on what a variable should mean (which is a problem with their chosen variable name). Alternatively can add an `override` option, which it can use to tell the consumer not to override if it already exists, but without guaranteed ordering it's a subpar approach compared to priorities. My biggest concern is monolith packages that provide resolvers, but don't allow targeted disabling. Currently, disabling a provider will have no effect; a restart would be required.

> curious about the case where those two packages try to add _conflicting_ resolvers, like if they both want to be in control of what `$FOO` resolves to

Currently, the latest to register gets it; it comes down to just setting a map key (variable name) to a value (resolver function).

(When I say "variable" in the above, I also mean transforms like `/upcase`)